### PR TITLE
Add memory error debugging description

### PR DESCRIPTION
**Pull request by negiyas**
Add the "Finding memory errors" section in docs/DebuggingNumericalError.md about how to use the '"Electric Fence library" for debugging.

### CODE REVIEW

File: docs/DebuggingNumericalError.md [hunk 1]
```diff
@@ -100,3 +100,27 @@ To print a message followed by one value, inject the following code (where `val`
...
+
+The "Electric Fence library" can be used for debugging these bugs. It helps
+you detect two common programming bugs: software that overruns the boundaries
+of a malloc() memory allocation, and software that touches a memory allocation
+that has been released by free(). Unlike other malloc() debuggers, Electric
+Fence will detect read accesses as well as writes, and it will pinpoint the
+exact instruction that causes an error.
+
+Because the library is not officially supported by RedHat, you need to download 
+the source code package (SRPM) from https://rpmfind.net/linux/RPM/fedora/devel/rawhide/s390x/e/ElectricFence-2.2.2-57.fc36.s390x.html,
```
[the remark this answers is not in the document]
Comment (negiyas) [reply docs/DebuggingNumericalError.md]: I updated the description. Thank for your comments.

[the remark this answers is not in the document]
Comment (tungld) [reply docs/DebuggingNumericalError.md]: It looks to me it is too detailed to list packages for different environments. There are many environments and we cannot list them all. For example, in the above list, packages are for fedora but not ubuntu, suse, etc. for linux, mac, but not windows.

Instead, we can suggest a list of tools that are useful to debug memory issues in onnx-mlir, e.g. valgrind, mtrace, efence. Which one is suggested to use in which situation, and some tips to use them in onnx-mlir, etc. Users should find and install the tools depending on their environments.

[the remark this answers is not in the document]
Comment (etiotto) [reply docs/DebuggingNumericalError.md]: > It looks to me it is too detailed to list packages for different environments. There are many environments and we cannot list them all. For example, in the above list, packages are for fedora but not ubuntu, suse, etc. for linux, mac, but not windows.

OK that is also a sensible solution. We should refrain from just listing a tool for s390x given that environment is just one of the possible environment for users of `onnx-mlir`.  List tools in a more generic way is fine by me.

[the remark this answers is not in the document]
Comment (negiyas) [reply docs/DebuggingNumericalError.md]: @etiotto @tungld  Thank you for your comments. I updated the description.
Your further comments and suggestions are very welcome.

File: docs/DebuggingNumericalError.md [hunk 2]
```diff
@@ -100,3 +100,29 @@ To print a message followed by one value, inject the following code (where `val`
...
+tool(https://github.com/sstefani/mtrace) are available. These tools trace memory
+allocation/free-related APIs, and can fix memory issues, such as memory leaks.
+
+However if the bugs relating to memory access, especially buffer overrun bugs,
+it is difficult to debug because run-time errors occur outside of the point
+ containing the bug. 
+The "Electric Fence library"(https://github.com/CheggEng/electric-fence) can be
+used for debugging these bugs. It helps you detect two common programming bugs: software that overruns the boundaries of a malloc() memory allocation, and
+software that touches a memory allocation
+that has been released by free(). Unlike other malloc() debuggers, Electric
```
Comment (etiotto) [inline docs/DebuggingNumericalError.md:118]: "malloc() debuggers" --> memory debuggers ?

**Review by etiotto — APPROVED** [review]
Thanks for the updated description @negiyas.

I think the information presented here is now in good shape, I left some suggestions on the grammar that  you can address before merging. 

LGTM

File: docs/DebuggingNumericalError.md [hunk 3]
```diff
@@ -100,3 +100,29 @@ To print a message followed by one value, inject the following code (where `val`
 ```code
 create.krnl.printf("inputElem: ", val, valType);
 ```
+
+## Finding memory errors
+
+If you know, or suspect, that onnx-mlir-compiled inference executable causes
```
[the remark this answers is not in the document]
Comment (negiyas) [reply docs/DebuggingNumericalError.md:106]: Thank you. Changed to "that an onnx-mlir compiled"

File: docs/DebuggingNumericalError.md [hunk 4]
```diff
@@ -100,3 +100,29 @@ To print a message followed by one value, inject the following code (where `val`
...
+bugs related to memory allcations. The valgrind
```
[the remark this answers is not in the document]
Comment (negiyas) [reply docs/DebuggingNumericalError.md:107]: Thank you!

File: docs/DebuggingNumericalError.md [hunk 5]
```diff
@@ -100,3 +100,29 @@ To print a message followed by one value, inject the following code (where `val`
...
+framework(https://valgrind.org/) or mtrace memory
+tool(https://github.com/sstefani/mtrace) are available. These tools trace memory
```
[the remark this answers is not in the document]
Comment (negiyas) [reply docs/DebuggingNumericalError.md:109]: done. Thank you!

File: docs/DebuggingNumericalError.md [hunk 6]
```diff
@@ -100,3 +100,29 @@ To print a message followed by one value, inject the following code (where `val`
...
+allocation/free-related APIs, and can fix memory issues, such as memory leaks.
```
[the remark this answers is not in the document]
Comment (negiyas) [reply docs/DebuggingNumericalError.md:110]: Done. Thank you!

File: docs/DebuggingNumericalError.md [hunk 7]
```diff
@@ -100,3 +100,29 @@ To print a message followed by one value, inject the following code (where `val`
...
+
+However if the bugs relating to memory access, especially buffer overrun bugs,
```
[the remark this answers is not in the document]
Comment (negiyas) [reply docs/DebuggingNumericalError.md:112]: Done. Thank you.
I also replace the word "bug" to the word "problem."

File: docs/DebuggingNumericalError.md [hunk 8]
```diff
@@ -100,3 +100,29 @@ To print a message followed by one value, inject the following code (where `val`
...
+it is difficult to debug because run-time errors occur outside of the point
```
[the remark this answers is not in the document]
Comment (negiyas) [reply docs/DebuggingNumericalError.md:113]: Done. Thank you.

File: docs/DebuggingNumericalError.md [hunk 9]
```diff
@@ -100,3 +100,29 @@ To print a message followed by one value, inject the following code (where `val`
...
+ containing the bug. 
+The "Electric Fence library"(https://github.com/CheggEng/electric-fence) can be
+used for debugging these bugs. It helps you detect two common programming bugs: software that overruns the boundaries of a malloc() memory allocation, and
```
[the remark this answers is not in the document]
Comment (negiyas) [reply docs/DebuggingNumericalError.md:116]: Done. Thank you.

File: docs/DebuggingNumericalError.md (hunk 2 above)
Comment (negiyas) [reply docs/DebuggingNumericalError.md:118]: Done. Thank you.

File: docs/DebuggingNumericalError.md [hunk 10]
```diff
@@ -100,3 +100,29 @@ To print a message followed by one value, inject the following code (where `val`
...
+software that touches a memory allocation
+that has been released by free(). Unlike other malloc() debuggers, Electric
+Fence will detect read accesses as well as writes, and it will pinpoint the
+exact instruction that causes an error.
+
+Since the Electric Fence library is not officially supported by RedHat, you
+need to download, build and install the source code by yourself to your system.
```
[the remark this answers is not in the document]
Comment (negiyas) [reply docs/DebuggingNumericalError.md:123]: Changed to "your system" to on yours"

File: docs/DebuggingNumericalError.md [hunk 11]
```diff
@@ -100,3 +100,28 @@ To print a message followed by one value, inject the following code (where `val`
 ```code
 create.krnl.printf("inputElem: ", val, valType);
 ```
+
+## Finding memory errors
+
+If you know, or suspect, that an onnx-mlir-compiled inference executable
+suffers from memory allocation related issues, the
+framework(https://valgrind.org/) or mtrace memory
```
Comment (tungld) [inline docs/DebuggingNumericalError.md:108]: Please use link syntax in markdown, e.g. `[framework](https://valgrind.org/)`. The same for the other urls.

**Review by tungld — APPROVED** [review]
LGTM with pending link syntax.

File: docs/DebuggingNumericalError.md (hunk 11 above)
Comment (negiyas) [reply docs/DebuggingNumericalError.md]: done. Thank you!

File: docs/DebuggingNumericalError.md [hunk 12]
```diff
@@ -100,3 +100,28 @@ To print a message followed by one value, inject the following code (where `val`
 ```code
 create.krnl.printf("inputElem: ", val, valType);
 ```
+
+## Finding memory errors
+
+If you know, or suspect, that an onnx-mlir-compiled inference executable
+suffers from memory allocation related issues, the
+[framework](https://valgrind.org/) or
```
Comment (tungld) [inline docs/DebuggingNumericalError.md:108]: "framework" -> "valgrind framework"?

File: docs/DebuggingNumericalError.md [hunk 13]
```diff
@@ -100,3 +100,29 @@ To print a message followed by one value, inject the following code (where `val`
...
+
+If you know, or suspect, that onnx-mlir-compiled inference executable causes
+bugs related to memory allcations. The valgrind
+framework(https://valgrind.org/) or mtrace memory
+tool(https://github.com/sstefani/mtrace) are available. These tools trace memory
+allocation/free-related APIs, and can fix memory issues, such as memory leaks.
+
+However if the bugs relating to memory access, especially buffer overrun bugs,
+it is difficult to debug because run-time errors occur outside of the point
+ containing the bug.
```
[the remark this answers is not in the document]
Comment (negiyas) [reply docs/DebuggingNumericalError.md]: done. Thank you.

File: docs/DebuggingNumericalError.md (hunk 12 above)
Comment (negiyas) [reply docs/DebuggingNumericalError.md:108]: done. Thank you.

**Pull request merged**